src/NuGet.Core/NuGet.Build.Tasks/NuGet.RestoreEx.targets

```diff
@@ -31,8 +31,15 @@ Copyright (c) .NET Foundation. All rights reserved.
         RestorePackagesConfig="$(RestorePackagesConfig)"
         SolutionPath="$(SolutionPath)"
         ProcessFileName="$(NuGetConsoleProcessFileName)"
-        MSBuildStartupDirectory="$(MSBuildStartupDirectory)"
-        />
+        MSBuildStartupDirectory="$(MSBuildStartupDirectory)">
+      <Output TaskParameter="RestoredProjectOutputPaths" ItemName="_RestoredProjectOutputPaths" />
+    </RestoreTaskEx>
+
+    <ItemGroup Condition="@(_RestoredProjectOutputPaths->Count()) > 0">
+      <EmbedInBinlog Include="%(_RestoredProjectOutputPaths.Identity)project.assets.json;
+                              %(_RestoredProjectOutputPaths.Identity)*.nuget.dgspec.json;
+                              %(_RestoredProjectOutputPaths.Identity)*.nuget.g.*" />
```

Comment:
Aren't these files already included as they are a part of the msbuild evaluation?

Reply:
I think yes, if they existed when msbuild was invoked. 
I believe this covers the scenario when they're created during restore, so this step injects those newly created files.

Reply:
Ah, great point!
They might be different and we'd benefit from having both.

Reply:
I think it's just the assets file that's included by the .NET SDK. But I think that when the .NET SDK adds the assets file to the binlog, it attaches the assets file before restore runs, so when the restore is not already up to date, it attaches the wrong data. The assets file is still very useful in a `--no-restore` build, so we probably want the .NET SDK to keep including the assets file on build (unless we add some targets on build to do it, so we can remove the logic from the .NET SDK itself).

Reply:
> I think it's just the assets file that's included by the .NET SDK

![image](https://user-images.githubusercontent.com/2878341/217629814-4b033eed-6445-49a2-ba51-ca8d8babbce7.png)

It's not the SDK, it's just the simple concept of msbuidl files that are part of the evaluation are included.